ibis/backends/tests/test_impure.py

```diff
@@ -0,0 +1,207 @@
+    pytest.mark.notyet(
+        "risingwave",
+        raises=PsycoPg2InternalError,
+        reason="function random() does not exist",
+    ),
+]
+
+no_udfs = [
+    pytest.mark.notyet("datafusion", raises=NotImplementedError),
+    pytest.mark.notimpl(
+        [
+            "bigquery",
+            "clickhouse",
+            "dask",
+            "druid",
+            "exasol",
+            "impala",
+            "mssql",
+            "mysql",
+            "oracle",
+            "pandas",
+            "trino",
+            "risingwave",
+        ]
+    ),
+    pytest.mark.notimpl("pyspark", reason="only supports pandas UDFs"),
+    pytest.mark.broken(
+        "flink",
+        condition=sys.version_info >= (3, 11),
+        raises=Py4JJavaError,
+        reason="Docker image has Python 3.10, results in `cloudpickle` version mismatch",
+    ),
+]
+
+no_uuids = [
+    pytest.mark.notimpl(
+        [
+            "druid",
+            "exasol",
+            "oracle",
+            "polars",
+            "pyspark",
+            "risingwave",
+            "pandas",
+            "dask",
+        ],
+        raises=com.OperationNotDefinedError,
+    ),
+    pytest.mark.broken("mssql", reason="Unrelated bug: Incorrect syntax near '('"),
+]
+
+
+@ibis.udf.scalar.python(side_effects=True)
+def my_random(x: float) -> float:
+    # need to make the whole UDF self-contained for postgres to work
+    import random
+
+    return random.random()  # noqa: S311
+
+
+mark_impures = pytest.mark.parametrize(
+    "impure",
+    [
+        pytest.param(
+            lambda _: ibis.random(),
+            marks=no_randoms,
+            id="random",
+        ),
+        pytest.param(
+            lambda _: ibis.uuid().cast(str).contains("a").ifelse(1, 0),
+            marks=[
+                *no_uuids,
+                pytest.mark.broken("impala", reason="instances are uncorrelated"),
+            ],
+            id="uuid",
+        ),
+        pytest.param(
+            lambda table: my_random(table.float_col),
+            marks=[
+                *no_udfs,
+                pytest.mark.broken(
+                    ["flink", "postgres"], reason="instances are uncorrelated"
+                ),
+            ],
+            id="udf",
+        ),
+    ],
+)
+
+
+@pytest.mark.broken("sqlite", reason="instances are uncorrelated")
+@mark_impures
+def test_impure_correlated(alltypes, impure):
+    # An "impure" expression is random(), uuid(), or some other non-deterministic UDF.
+    # If we evaluate it for two different rows in the same relation,
+    # we might get different results. This is expected.
+    # But, as soon as we .select() it into a new relation, then that "locks in" the
+    # value, and any further references to it will be the same.
+    # eg if you look at the following SQL:
+    # WITH
+    #   t AS (SELECT random() AS common)
+    # SELECT common as x, common as y FROM t
+    # Then both x and y should have the same value.
```

Comment:
I don't think this is actually an expectation that Ibis can enforce.

Trino, for example evaluates CTEs once **per reference**, leading to annoying behavior like this:

```
trino:default> with t as (select 1 as i, random() as x) select * from t union select * from t;
 i |         x
---+--------------------
 1 | 0.8021965004733393
 1 | 0.3296580934824612
(2 rows)

Query 20240724_211040_00063_56j8j, FINISHED, 1 node
Splits: 98 total, 98 done (100.00%)
0.22 [0 rows, 0B] [0 rows/s, 0B/s]
```

[This issue has been open for years](https://github.com/trinodb/trino/issues/10), it's issue 10 on an issue tracker with nearly 23k total issues + PRs combined, for a project that is used in production at multiple large companies, so I don't have high hopes that it will be addressed any time soon.

No amount of CTE-ization that we do is going to change the way Trino works.

Reply:
Hmm, darn, I agree with your analysis of the situation. I agree it sounds like we can never guarantee this for trino. 

But that is trinos fault for not following the SQL standard, and I don't think we should throw out "better" semantics because they're not "perfect". Ibis exposes plenty of APIs that are inconsistent between backends, but those are much less obvious about what the proper behavior should be. If we feel OK about exposing inconsistent behavior in those circumstances, when it is not obvious what they correct behavior is, then I think it should be an easy decision to expose this inconsistent behavior.

Any user of trino who runs into this bug is out in the cold no matter what we decide here, they're gonna get weird results. But I as a user of duckdb or the other backends could see benefits if we make it so each select/mutate is a new relation.

If we don't enforce the semantics of "one select=one relation", then what is the alternative here? Say I wanted to end up with the above SQL, so that I could call random() once and then reference it multiple times. Is ibis making no guarantees, ie the only way would be for me to cache the intermediate result ie `x.mutate(common=ibis.random()).cache().select(x="common", y="common")`?

Reply:
wait, am I reading these test markers right that trino is actually passing this test? If I try

```python
import ibis
e = ibis.table({"x": int}).select(common=ibis.random()).select(x="common", y="common")
ibis.to_sql(e, "trino")
```

I get 

```sql
SELECT
  "t1"."common" AS "x",
  "t1"."common" AS "y"
FROM (
  SELECT
    RAND() AS "common"
  FROM "unbound_table_1" AS "t0"
) AS "t1"
```

which looks to me like it should be susceptible to the bug you link above?

Reply:
It might pass, but it's not **guaranteed** to.